lang/cs/Org.Apache.REEF.IMRU/OnREEF/CheckpointHandler/IMRUCheckpointHandler.cs

```diff
@@ -0,0 +1,128 @@
+        private IMRUCheckpointHandler(
+            [Parameter(typeof(CheckpointFilePath))] string checkpointFilePath,
+            IFileSystem fileSystem)
+        {
+            _fileSystem = fileSystem;
+            _checkpointFilePath = checkpointFilePath;
+
+            Logger.Log(Level.Info, "State file path: {0}", checkpointFilePath);
+        }
+
+        /// <summary>
+        /// Save serialized checkpoint data to remote checkpoint file.
+        /// </summary>
+        /// <param name="taskState"></param>
+        /// <param name="codec"></param>
+        public void Persistent(ITaskState taskState, ICodec<ITaskState> codec)
+        {
+            var localStateFile = Path.GetTempPath() + Guid.NewGuid().ToString("N").Substring(0, 4);
+            var localFlagfile = Path.GetTempPath() + Guid.NewGuid().ToString("N").Substring(0, 4);
+
+            string tick = DateTime.Now.Ticks.ToString();
+            string stateFileDir = Path.Combine(_checkpointFilePath, StateDir + tick);
+            string remoteStateFileName = Path.Combine(stateFileDir, StateFile + tick + StateFileExt);
+            string remoteFlagFileName = Path.Combine(stateFileDir, FlagFile + tick + FlagFileExt);
+
+            var stateFileUri = _fileSystem.CreateUriForPath(remoteStateFileName);
+            var flagFileUri = _fileSystem.CreateUriForPath(remoteFlagFileName);
+
+            var data = codec.Encode(taskState);
+            File.WriteAllBytes(localStateFile, data);
+            File.WriteAllText(localFlagfile, remoteStateFileName);
+
+            _fileSystem.CopyFromLocal(localStateFile, stateFileUri);
+            _fileSystem.CopyFromLocal(localFlagfile, flagFileUri);
+
+            File.Delete(localStateFile);
+            File.Delete(localFlagfile);
+        }
+
+        /// <summary>
+        /// Read checkpoint data and deserialize it into ITaskState object.
+        /// </summary>
+        /// <param name="codec"></param>
+        /// <returns></returns>
+        public ITaskState Restore(ICodec<ITaskState> codec)
+        {
+            if (!string.IsNullOrEmpty(_checkpointFilePath))
+            {
+                var files = _fileSystem.GetChildren(_fileSystem.CreateUriForPath(_checkpointFilePath));
+                if (files != null)
+                {
+                    var flagFiles = files.Where(f => f.AbsolutePath.Contains(FlagFile));
+                    var uris = flagFiles.OrderByDescending(ff => _fileSystem.GetFileStatus(ff).ModificationTime);
+
+                    Uri latestFlagFile = uris.FirstOrDefault();
+                    if (latestFlagFile != null)
+                    {
+                        var localLatestFlagfile = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N").Substring(0, 4));
+                        var localLatestStatefile = Path.Combine(Path.GetTempPath() + Guid.NewGuid().ToString("N").Substring(0, 4));
+
+                        _fileSystem.CopyToLocal(latestFlagFile, localLatestFlagfile);
```

Comment:
The copy to local should not be necessary.  Can't you just `.Open()` the remote file?

Reply:
Yeah, we can read directly from remote. I just want to match with the way the data is written to ensure the sate format.